model/context.go

```diff
@@ -55,10 +58,67 @@ type Url struct {
 	Fragment *string
 }
 
-// Page consists of Url string and referer
+func ParseURL(original, hostname string) *URL {
+	original = truncate(original)
+	url, err := url.Parse(original)
+	if err != nil {
+		return &URL{Original: &original}
+	}
+	if url.Scheme == "" {
+		url.Scheme = "http"
+	}
+	if url.Host == "" {
+		url.Host = hostname
+	}
+	full := truncate(url.String())
+	out := &URL{
+		Original: &original,
+		Scheme:   &url.Scheme,
+		Full:     &full,
+	}
+	if path := truncate(url.Path); path != "" {
+		out.Path = &path
+	}
+	if query := truncate(url.RawQuery); query != "" {
+		out.Query = &query
+	}
+	if fragment := url.Fragment; fragment != "" {
+		out.Fragment = &fragment
+	}
+	host, port, err := net.SplitHostPort(url.Host)
+	if err != nil {
+		host = truncate(url.Host)
+		port = ""
+	}
+	if host = truncate(host); host != "" {
+		out.Domain = &host
+	}
+	if port = truncate(port); port != "" {
+		if intv, err := strconv.Atoi(port); err == nil {
+			out.Port = &intv
+		}
+	}
+	return out
+}
+
+// truncate returns s truncated at n runes, and the number of runes in the resulting string (<= n).
+func truncate(s string) string {
+	var j int
+	for i := range s {
+		if j == 1024 {
+			return s[:i]
+		}
+		j++
+	}
+	return s
+}
+
+// Page consists of URL and referer
 type Page struct {
-	Url     *string
-	Referer *string
+	URL *URL
+	// Duplicated for BC reasons, remove in 8.0
+	URLString *string
```

Comment:
Can we delete URLString and use URL.Original? IIANM it's always exactly the same value.